### PR TITLE
User refactoring

### PR DESCRIPTION
**Pull request by ritosilva**
Start review on this code

### CODE REVIEW

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java [hunk 1]
```diff
@@ -115,6 +123,24 @@ public void deactivateCourseExecution(int executionId) {
...
 
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void anonymizeCourseExecutionUsers(int executionId) {
+        CourseExecution courseExecution = courseExecutionRepository.findById(executionId).orElseThrow(() -> new TutorException(COURSE_EXECUTION_NOT_FOUND));
+        for (User user : courseExecution.getUsers()) {
+            String oldUsername = user.getUsername();
+            authUserRepository.delete(user.getAuthUser());
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:134]: do user.getAuthUser().remove() first

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java [hunk 2]
```diff
@@ -115,6 +123,24 @@ public void deactivateCourseExecution(int executionId) {
...
+            user.setAuthUser(null);
+            String newUsername = user.getUsername();
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:136]: Change the username in the db, do not redefine the getUsername, we want to anonymize even to the database admin

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java [hunk 3]
```diff
@@ -174,7 +201,7 @@ public CourseExecution getDemoCourseExecution() {
     public List<ExternalUserDto> getExternalUsers(Integer courseExecutionId){
         CourseExecution execution = getExternalCourseExecution(courseExecutionId);
         return execution.getStudents().stream()
-                .sorted(Comparator.comparing(User::getUsername))
+                .sorted(Comparator.comparing(User::getName))
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:204]: Why not username?

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java [hunk 4]
```diff
@@ -174,7 +201,7 @@ public CourseExecution getDemoCourseExecution() {
...
                 .map(ExternalUserDto::new)
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:205]: Do the sort here when you have all info

Comment (ritosilva) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:205]: But we need to discuss all the cases where user.getAuthUser() is null...

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/impexp/domain/UsersXmlExport.java [hunk 5]
```diff
@@ -71,4 +68,42 @@ private void exportUserCourseExecutions(Element userElement, Set<CourseExecution
...
 	}
+
+	private void exportAuthUsers(Element userElement, AuthUser authUser) {
+		Element authUsersElement = new Element("authUsers");
+		Element authUserElement = new Element("authUser");
+
+		authUserElement.setAttribute("username", authUser.getUsername());
+		authUserElement.setAttribute("email", authUser.getEmail());
+
+		if (authUser.getType() != null) {
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/impexp/domain/UsersXmlExport.java:79]: Do we need this? is it possible to have AuthUser with getType() null?

**Review by ritosilva — COMMENTED** [review]
backend partial

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/AuthUserService.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public ExternalUserDto confirmRegistrationTransactional(ExternalUserDto externalUserDto) {
+        AuthExternalUser authUser = (AuthExternalUser) this.findAuthUserByUsername(externalUserDto.getUsername());
+
+        if (authUser == null) {
+            throw new TutorException(EXTERNAL_USER_NOT_FOUND, externalUserDto.getUsername());
+        }
+        if (authUser.isActive()) {
```
[the remark this answers is not in the document]
Comment (mgoulao) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/AuthUserService.java:227]: Null check or isActive check?

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java (hunk 2 above)
Comment (mgoulao) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:136]: Username is an AuthUser's property and without an object (`user.setAuthUser(null);` `orphanRemoval = true`), there is nothing to write.

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/FenixEduInterface.java [hunk 7]
```diff
@@ -65,27 +68,52 @@ private JsonObject getPersonCourses() {
...
+                                        .get("evaluationPeriod").getAsJsonObject()
+                                        .get("end").getAsString();    
+            if (!endDate.isEmpty()) {
+                LocalDateTime evaluationEnd = LocalDateTime.parse(endDate, formatter);
+                if (lastDate == null || evaluationEnd.isAfter(lastDate)) {
+                    lastDate = evaluationEnd;
+                }
+            }             
+        }
+        return DateHandler.toISOString(lastDate);
```
[the remark this answers is not in the document]
Comment (mgoulao) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/FenixEduInterface.java:101]: Yes, if no information about evaluations is found the endDate is set to null. Frontend when determining if the course execution has ended: 
``` 
  hasCourseSemesterFinished(course: Course):boolean {
    if(course.endDate) {
      return new Date(course.endDate) < new Date();
    } else if (course.academicTerm
               && RegExp(/[1-2]º?\s?\w+\s[0-9]+\/[0-9]+/).test(course.academicTerm)) {
      const termTokens = course.academicTerm.split(/º|\s|\//);
      const month = termTokens[0] ===  "1" ? 3 : 9; // march : september
      const year = termTokens[termTokens.length-1]
      return new Date(`${year}-${month}-${1}`) < new Date();
    }
    return false;
  }
```

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/FenixEduInterface.java [hunk 8]
```diff
@@ -65,27 +68,52 @@ private JsonObject getPersonCourses() {
...
+    private List<CourseDto> getCourses(String type) {
+        JsonArray coursesJson = courses.get(type).getAsJsonArray();
         List<CourseDto> result = new ArrayList<>();
         for (JsonElement courseJson : coursesJson) {
-            result.add(new CourseDto(courseJson.getAsJsonObject().get("name").getAsString(),
+            CourseDto course = new CourseDto(courseJson.getAsJsonObject().get("name").getAsString(),
                     courseJson.getAsJsonObject().get("acronym").getAsString(),
-                    courseJson.getAsJsonObject().get("academicTerm").getAsString()));
+                    courseJson.getAsJsonObject().get("academicTerm").getAsString());
+            if(type.equals("teaching")) {
```
[the remark this answers is not in the document]
Comment (ritosilva) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/FenixEduInterface.java]: format, space between if and (

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/AuthUserService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ritosilva) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/AuthUserService.java:226]: Inside confirmRegistration, is referred in comment below

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java (hunk 2 above)
Comment (ritosilva) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:138]: OK!

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java [hunk 9]
```diff
@@ -115,6 +123,26 @@ public void deactivateCourseExecution(int executionId) {
...
     }
 
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void anonymizeCourseExecutionUsers(int executionId) {
+        CourseExecution courseExecution = courseExecutionRepository.findById(executionId).orElseThrow(() -> new TutorException(COURSE_EXECUTION_NOT_FOUND));
+        for (User user : courseExecution.getUsers()) {
+            if (user.getAuthUser() == null) continue;
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/course/CourseService.java:133]: if (user.getUser() != null) {

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/AuthUserService.java [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    public AuthDto fenixAuth(FenixEduInterface fenix) {
+        String username = fenix.getPersonUsername();
+        List<CourseDto> fenixAttendingCourses = fenix.getPersonAttendingCourses();
+        List<CourseDto> fenixTeachingCourses = fenix.getPersonTeachingCourses();
+
+        List<CourseExecution> activeAttendingCourses = getActiveTecnicoCourses(fenixAttendingCourses);
+        List<CourseExecution> activeTeachingCourses = getActiveTecnicoCourses(fenixTeachingCourses);
+        AuthTecnicoUser authUser;
+        try {
+            authUser = (AuthTecnicoUser) this.findAuthUserByUsername(username);
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/auth/AuthUserService.java:73]: remove this.

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserServiceApplicational.java [hunk 11]
```diff
@@ -23,33 +26,38 @@
...
         return user;
     }
 
     public NotificationResponse<CourseDto> importListOfUsers(InputStream stream, int courseExecutionId) {
         NotificationResponse<CourseDto> courseDtoNotificationResponse = userService.importListOfUsersTransactional(stream, courseExecutionId);
         courseDtoNotificationResponse.getResponse().getCourseExecutionUsers()
-                .forEach(this::sendConfirmationEmailTo);
+                .stream()
+                .map(UserDto::getUsername)
+                .map(authUserRepository::findAuthUserByUsername)
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserServiceApplicational.java:49]: Why doing this? It is possible to trust userService.importListOfUsersTransactional

**Review by ritosilva — COMMENTED** [review]
backend review, except tests

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/course/service/AnonymizeCourseExecutionUsersTest.groovy [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        user.name == "Student " + user.id
+        user.getUsername() == "Student-" + user.id
+        answerItemsIds.equals(
+                questionAnswerItemRepository.findQuestionAnswerItemsByUsername(user.getUsername())
+                .stream().map({ answerItem -> answerItem.getId() })
+                .collect(Collectors.toList())
+        )
+    }
+
+
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/course/service/AnonymizeCourseExecutionUsersTest.groovy:58]: remove lines

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/impexp/service/ImportExportUsersTest.groovy [hunk 13]
```diff
@@ -5,27 +5,96 @@ import org.springframework.boot.test.context.TestConfiguration
...
+        and: 'a clean database'
+        userRepository.deleteAll()
+        externalCourseExecution.getUsers().clear()
+
+        when:
+        userService.importUsers(usersXml)
+
+        then:
+        externalCourseExecution.getUsers().size() == 2
+
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/impexp/service/ImportExportUsersTest.groovy:52]: and:

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/impexp/service/ImportExportUsersTest.groovy [hunk 14]
```diff
@@ -5,27 +5,96 @@ import org.springframework.boot.test.context.TestConfiguration
...
+        userOne.getRole() == User.Role.TEACHER
+        userOne.getCourseExecutions().size() == 1
+        userOne.getAuthUser().getEmail() == USER_1_EMAIL
+        userOne.getAuthUser().getUsername() == USER_1_USERNAME
+        userOne.getAuthUser().getPassword() == USER_1_PASSWORD
+        userOne.getAuthUser().getLastAccess() == LOCAL_DATE_TODAY
+        userOne.getAuthUser().getConfirmationToken() == USER_1_TOKEN
+        userOne.getAuthUser().getTokenGenerationDate() == LOCAL_DATE_TODAY
+        userOne.getAuthUser().isActive()
+
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/impexp/service/ImportExportUsersTest.groovy:67]: and:

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/submission/webservice/ToggleInReviewStatusWebServiceIT.groovy [hunk 15]
```diff
@@ -56,8 +58,9 @@ class ToggleInReviewStatusWebServiceIT extends SpockTest {
         questionSubmissionService.createQuestionSubmission(questionSubmissionDto)
         questionSubmission = questionSubmissionRepository.findAll().get(0)
 
-        teacher = new User(USER_2_NAME, USER_2_EMAIL, USER_2_EMAIL, User.Role.TEACHER, true, false)
-        teacher.setPassword(passwordEncoder.encode(USER_2_PASSWORD))
+        teacher = new User(USER_2_NAME, USER_2_EMAIL, USER_2_EMAIL,
+                User.Role.TEACHER, false, AuthUser.Type.TECNICO)
+        (teacher.authUser).setPassword(passwordEncoder.encode(USER_2_PASSWORD))
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/submission/webservice/ToggleInReviewStatusWebServiceIT.groovy:63]: ???

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
+import org.springframework.boot.test.context.TestConfiguration
+import pt.ulisboa.tecnico.socialsoftware.tutor.BeanConfiguration;
+import pt.ulisboa.tecnico.socialsoftware.tutor.SpockTest
+import pt.ulisboa.tecnico.socialsoftware.tutor.user.User
+
+@DataJpaTest
+class CreateAuthUserTest  extends SpockTest {
+
+    User user
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy:12]: def user

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    AuthUser authUser
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy:13]: def  AuthUser

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        authUserRepository.save(authUser)
+
+        when:
+        authUser = authUserRepository.findAuthUserByUsername(USER_1_USERNAME).get()
+
+        then:
+        authUser.username == USER_1_USERNAME
+        authUser.email == USER_1_EMAIL
+        authUser.isActive()
+        authUser instanceof AuthTecnicoUser
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy:32]: check that is connected with the user

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+        authUserRepository.save(authUser)
+
+        when:
+        authUser = authUserRepository.findAuthUserByUsername(USER_1_USERNAME).get()
+
+        then:
+        authUser.username == USER_1_USERNAME
+        authUser.email == USER_1_EMAIL
+        !authUser.isActive()
+        authUser instanceof AuthExternalUser
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy:46]: idem

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+        authUserRepository.save(authUser)
+
+        when:
+        authUser = authUserRepository.findAuthUserByUsername(USER_1_USERNAME).get()
+
+        then:
+        authUser.username == USER_1_USERNAME
+        authUser.email == USER_1_EMAIL
+        authUser.isActive()
+        authUser instanceof AuthDemoUser
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy:60]: idem

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+    User user
+    AuthUser authUser
+
+    def setup() {
+        user = new User(USER_1_NAME, User.Role.STUDENT, false)
+        userRepository.save(user)
+        user.setKey(user.getId())
+    }
+
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateAuthUserTest.groovy:20]: These 3 tests can be in a table

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy [hunk 22]
```diff
@@ -22,10 +22,11 @@ import spock.lang.Unroll;
 @DataJpaTest
 class CreateUserTest extends SpockTest {
     def user
+
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy:25]: This class containts tests of several methods, it should be split

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy [hunk 23]
```diff
@@ -64,51 +65,85 @@ class CreateUserTest extends SpockTest {
...
-        result.isActive() == false
         !result.isAdmin()
+        result.getAuthUser() != null
+        result.getAuthUser().getUsername() == USER_1_USERNAME
+        result.getAuthUser().getEmail() == USER_1_EMAIL
+        result.getAuthUser() instanceof AuthDemoUser
     }
 
+
     def "checkConfirmationToken: correct token and date has not expired" (){
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy:128]: different class

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy [hunk 24]
```diff
@@ -239,11 +274,11 @@ class CreateUserTest extends SpockTest {
         user.getQuizAnswers().size() == previousNumberQuizAnswers + 1
     }
 
-    def "remove from course executions" (){
+    def "remove inactive user from course executions" (){
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy:277]: It is necessary to check that the user does not exist
It is necessary to check that the authUser is also removed

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy [hunk 25]
```diff
@@ -255,14 +290,35 @@ class CreateUserTest extends SpockTest {
         courseExecution.getUsers().size() == previousNumberOfUsers - 1
     }
 
+    def "remove active user from course executions" (){
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/domain/CreateUserTest.groovy:293]: idem
and it is nor from course execution, is the remove of a user....

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/service/ConfirmRegistrationTest.groovy [hunk 26]
```diff
@@ -65,10 +66,10 @@ class ConfirmRegistrationTest extends SpockTest {
     def "user is already active" () {
         given: "an active user"
         externalUserDto.setPassword(USER_1_PASSWORD)
-        authServiceApplicational.confirmRegistration(externalUserDto)
+        authUserServiceApplicational.confirmRegistration(externalUserDto)
 
         when:
-        authServiceApplicational.confirmRegistration(externalUserDto)
+        authUserServiceApplicational.confirmRegistration(externalUserDto)
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/service/ConfirmRegistrationTest.groovy:72]: this is not good to use the test to be tested in the setup...

File: backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/service/ResetDemoStudentTest.groovy [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+    def setup(){
+        course = new Course(COURSE_1_NAME, Course.Type.EXTERNAL)
+        courseRepository.save(course)
+        courseExecution = new CourseExecution(course, COURSE_1_ACRONYM, COURSE_1_ACADEMIC_TERM, Course.Type.EXTERNAL, LOCAL_DATE_TOMORROW)
+        courseExecutionRepository.save(courseExecution)
+    }
+
+    def "reset all demo students"() {
+        given:
+        def user1 = userService.createDemoStudent()
```
Comment (ritosilva) [inline backend/src/test/groovy/pt/ulisboa/tecnico/socialsoftware/tutor/user/service/ResetDemoStudentTest.groovy:30]: do not use service to create the user

**Review by ritosilva — COMMENTED** [review]
Changes

File: backend/src/main/resources/application-dev.properties.example [hunk 28]
```diff
@@ -1,21 +0,0 @@
-## Datasource
```
[the remark this answers is not in the document]
Comment (ritosilva) [reply backend/src/main/resources/application-dev.properties.example]: .example not .example.txt

File: backend/src/main/resources/application-prod.properties.example [hunk 29]
```diff
@@ -1,18 +0,0 @@
-## Datasource
```
[the remark this answers is not in the document]
Comment (ritosilva) [reply backend/src/main/resources/application-prod.properties.example]: idem

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserService.java [hunk 30]
```diff
@@ -244,16 +232,16 @@ public void resetDemoStudents() {
             propagation = Propagation.REQUIRED)
     public ExternalUserDto createExternalUserTransactional(Integer courseExecutionId, ExternalUserDto externalUserDto) {
         CourseExecution courseExecution = getExternalCourseExecution(courseExecutionId);
-        User user = getOrCreateUser(externalUserDto);
+        User user = getOrCreateUser(externalUserDto).getUser();
         associateUserWithExecution(courseExecution, user);
-        generateConfirmationToken(user);
-        return new ExternalUserDto(user);
+        generateConfirmationToken(user.getAuthUser());
```
[the remark this answers is not in the document]
Comment (ritosilva) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserService.java]: authUser.generateConfirmationToken()

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserService.java [hunk 31]
```diff
@@ -244,16 +232,16 @@ public void resetDemoStudents() {
             propagation = Propagation.REQUIRED)
     public ExternalUserDto createExternalUserTransactional(Integer courseExecutionId, ExternalUserDto externalUserDto) {
         CourseExecution courseExecution = getExternalCourseExecution(courseExecutionId);
-        User user = getOrCreateUser(externalUserDto);
+        User user = getOrCreateUser(externalUserDto).getUser();
         associateUserWithExecution(courseExecution, user);
```
[the remark this answers is not in the document]
Comment (ritosilva) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserService.java]: invoke user.addCourseExecution(...) and implement this inside the class is more object-oriented

**Pull request merged**